test/relayserver/RegistrationManager.test.ts

```diff
@@ -0,0 +1,724 @@
+      registrationManager._extractDuePendingEvents(2);
+
+      expect(registrationManager._delayedEvents).to.be.deep.equal([
+        { block: 3, eventData: {} as TypedEvent },
+      ]);
+    });
+  });
+
+  describe('attemptRegistration', function () {
+    let relayServer: RelayServer;
+    let registrationManager: RegistrationManager;
+
+    beforeEach(async function () {
+      const relayOwner = (await ethers.getSigners()).at(0) as SignerWithAddress;
+      const relayHub = await deployRelayHub(undefined, {
+        maxWorkerCount,
+        minimumEntryDepositValue,
+        minimumUnstakeDelay,
+        minimumStake,
+      });
+      loadConfiguration({
+        app: basicAppConfig,
+        contracts: {
+          relayHubAddress: relayHub.address,
+        },
+      });
+      relayServer = await getFundedServer(
+        { relayOwner },
+        undefined,
+        unstakeDelay
+      );
+      await relayServer.init();
+      registrationManager = relayServer.registrationManager;
+      registrationManager.balanceRequired.requiredValue = constants.Zero;
+      registrationManager.stakeRequired.requiredValue = constants.Zero;
+      await registrationManager.refreshStake();
+    });
+
+    afterEach(function () {
+      config.util.extendDeep(config, originalConfig);
+    });
+
+    describe('without re-registration', function () {
+      it('should register server and add workers', async function () {
+        let pendingTransactions = await relayServer.txStoreManager.getAll();
+
+        expect(pendingTransactions.length).to.be.equal(0);
+
+        const receipts = await registrationManager.attemptRegistration(1);
+
+        await assertEventHub('RelayServerRegistered', receipts);
+        await assertEventHub('RelayWorkersAdded', receipts);
+
+        pendingTransactions = await relayServer.txStoreManager.getAll();
+
+        const addWorkerTransaction =
+          pendingTransactions[0] as StoredTransaction;
+        const registerServerTransaction =
+          pendingTransactions[1] as StoredTransaction;
+
+        expect(pendingTransactions.length).to.be.gt(0);
+        expect(addWorkerTransaction.serverAction).to.be.equal(
+          ServerAction.ADD_WORKER
+        );
+        expect(registerServerTransaction.serverAction).to.be.equal(
+          ServerAction.REGISTER_SERVER
+        );
+      });
+    });
+  });
+
+  describe('assertRegistered', function () {
+    let relayServer: RelayServer;
+
+    beforeEach(async function () {
+      const relayOwner = (await ethers.getSigners()).at(0) as SignerWithAddress;
+      const relayHub = await deployRelayHub();
+      loadConfiguration({
+        app: basicAppConfig,
+        contracts: {
+          relayHubAddress: relayHub.address,
+        },
+      });
+      relayServer = await getInitiatedServer({ relayOwner });
+    });
+
+    afterEach(function () {
+      config.util.extendDeep(config, originalConfig);
+    });
```

Comment:
why don't we move this afterEach to an upper level? If so, we wouldn't need to repeat it each time

Reply:
I just keep it for structure, but there are some  afterEach that are not really necessary. I just have it like that to try to be more realistic.

Reply:
updated